### ISSUE TITLE
Real-time prediction delay issue

### ISSUE THREAD

**Issue by ghiandonigianmarco** [scikit-multilearn/scikit-multilearn#157]
Hello everyone,

I am testing both BR and CC combined with Random Forests by training them with a dataset of shape 340.000 rows, 167 features, and 319 classes. Once the model is trained and loaded in memory, I need to use it to produce real-time predictions but it takes a fixed time of ~30 seconds to wake up, independently from the number of test rows I ask the predictions for. A summary of the times for increasing magnitudes of entries is reported as follows for CC-RF (BR-RF reported similar results):

Entries | Time (ms)
- 1 : 33,050
- 10 : 33,060
- 100 : 33,120
- 1000 : 33,920
- 10000 : 42,240

The summary shows that there is a constant amount of time necessary to start the prediction process and going back to the settings I have used I could not find any data processing method that demanding. I have used these general settings just to make a bit of context:

```
rf = RandomForestClassifier(n_estimators=10, criterion='gini', max_depth=None, 
                            min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, 
                            max_features='auto', max_leaf_nodes=None, min_impurity_decrease=0.0, 
                            min_impurity_split=None, bootstrap=True, oob_score=False, n_jobs=4, 
                            random_state=11, verbose=0, warm_start=False, class_weight=None)
br = BinaryRelevance(classifier=rf, require_dense=[True,True])
cc = ClassifierChain(classifier=clf, require_dense=[True,True])
```

Unfortunately, I can share neither training nor test set with you but I suspect that this could be a generic issue that can be reproduced with any set. Furthermore, I have trained a model in WEKA directly in Java (which takes ageeees to train, dump, and load though) then tested it and it takes just a few milliseconds per prediction.

Does anyone have an idea of why this takes so much time to start the prediction?

Best,

Giammy

**Comment by Rami-Majdoub** [comment]
changing the features **X** from a sparse to dense is not needed!
because the **X** can be a sparse matrix but the **y** can't.
you need to change your code to this : **require_dense=[False,True]**

documentation for RandomForestClassifier : [link](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier)
   >  fit(X, y, sample_weight=None)
        X : array-like **or sparse matrix** of shape = [n_samples, n_features]
        y : array-like, shape = [n_samples] or [n_samples, n_outputs]

this will hopefully :pray:  decrease the training time and the memory usage
